### PR TITLE
Instrument obsreport.Scraper

### PR DESCRIPTION
**Pull request by moh-osman3**
Description:

This PR instruments obsreport.Scraper with otel go. This is a followup PR that is based on similar changes made for obsreport.Receiver: https://github.com/open-telemetry/opentelemetry-collector/pull/6222

Link to tracking Issue: Part of https://github.com/open-telemetry/opentelemetry-collector/issues/816

Testing:
Ran contrib collector with prometheusreceiver and hostmetricsreceiver and confirmed that new metrics show up. Tested both with feature gate disabled and enabled.

```
curl http://localhost:8888/metrics | tail -n 13
# HELP otelcol_scraper_errored_metric_points_ratio_total Number of metric points that were unable to be scraped.
# TYPE otelcol_scraper_errored_metric_points_ratio_total counter
otelcol_scraper_errored_metric_points_ratio_total{receiver="hostmetrics",scraper="memory"} 0
# HELP otelcol_scraper_errored_metric_points_ratio_total_total Number of metric points that were unable to be scraped.
# TYPE otelcol_scraper_errored_metric_points_ratio_total_total counter
otelcol_scraper_errored_metric_points_ratio_total_total{receiver="hostmetrics",scraper="cpu"} 4
# HELP otelcol_scraper_scraped_metric_points_ratio_total Number of metric points successfully scraped.
# TYPE otelcol_scraper_scraped_metric_points_ratio_total counter
otelcol_scraper_scraped_metric_points_ratio_total{receiver="hostmetrics",scraper="cpu"} 0
# HELP otelcol_scraper_scraped_metric_points_ratio_total_total Number of metric points successfully scraped.
# TYPE otelcol_scraper_scraped_metric_points_ratio_total_total counter
otelcol_scraper_scraped_metric_points_ratio_total_total{receiver="hostmetrics",scraper="memory"} 2
# HELP otelcol_target_info Target metadata
```

### CODE REVIEW

File: obsreport/obsreporttest/otelprometheuschecker.go [hunk 1]
```diff
@@ -34,6 +34,13 @@ type prometheusChecker struct {
 	promHandler http.Handler
 }
 
+func (pc *prometheusChecker) checkScraperMetrics(receiver config.ComponentID, scraper config.ComponentID, scrapedMetricPoints, erroredMetricPoints int64) error{
```
Comment (paivagustavo) [inline obsreport/obsreporttest/otelprometheuschecker.go:37]: Please rebase and add tests for these as well.

File: obsreport/obsreport_scraper.go [hunk 2]
```diff
@@ -47,19 +66,72 @@ type ScraperSettings struct {
...
+func NewScraper(cfg ScraperSettings) *Scraper {
+	scr, err := newScraper(cfg, featuregate.GetRegistry())
+	if err != nil && cfg.ReceiverCreateSettings.Logger != nil {
+		cfg.ReceiverCreateSettings.Logger.Warn("Error creating an obsreport.Scraper", zap.Error(err))
+	}
+	return scr
+}
 
-// MustNewScraper creates a new Scraper.
+// MustNewReceiver creates a new Scraper.
```
Comment (paivagustavo) [inline obsreport/obsreport_scraper.go:77]:
```suggestion
// MustNewScraper creates a new Scraper.
```

File: obsreport/obsreporttest/otelprometheuschecker.go (hunk 1 above)
Comment (moh-osman3) [reply obsreport/obsreporttest/otelprometheuschecker.go:37]: Thanks for the review! Done!

File: obsreport/obsreport_scraper.go [hunk 3]
```diff
@@ -46,19 +65,72 @@ type ScraperSettings struct {
...
 
 // Deprecated: [v0.64.0] use MustNewScraper.
-var NewScraper = MustNewScraper
+func NewScraper(cfg ScraperSettings) *Scraper {
+	scr, err := newScraper(cfg, featuregate.GetRegistry())
+	if err != nil && cfg.ReceiverCreateSettings.Logger != nil {
+		cfg.ReceiverCreateSettings.Logger.Warn("Error creating an obsreport.Scraper", zap.Error(err))
+	}
+	return scr
+}
```
Comment (bogdandrutu) [inline obsreport/obsreport_scraper.go:74]: Already changed, https://github.com/open-telemetry/opentelemetry-collector/blob/main/obsreport/obsreport_scraper.go#L49, you need to rebase this.

Comment (moh-osman3) [reply obsreport/obsreport_scraper.go]: Thanks for the review! Rebased so things should be up to date now.

File: .chloggen/obsreport-scraper.yaml [hunk 4]
```diff
@@ -0,0 +1,17 @@
+
```
Comment (bogdandrutu) [inline .chloggen/obsreport-scraper.yaml:1]:
```suggestion
```

File: obsreport/obsreport_scraper.go [hunk 5]
```diff
@@ -46,26 +65,72 @@ type ScraperSettings struct {
 }
 
 // NewScraper creates a new Scraper.
-func NewScraper(cfg ScraperSettings) (*Scraper, error) {
-	return &Scraper{
+func NewScraper(cfg ScraperSettings) *Scraper {
```
Comment (bogdandrutu) [inline obsreport/obsreport_scraper.go:68]: Please don't break this API. Return error, do not log.

Comment (moh-osman3) [reply obsreport/obsreport_scraper.go]: Ah sorry I'm scatterbrained this week, that was a careless mistake.  Fixed this now and retested this change. Thanks for catching this!

File: obsreport/obsreport_scraper.go [hunk 6]
```diff
@@ -118,3 +176,15 @@ func (s *Scraper) EndMetricsOp(
...
+	if s.useOtelForMetrics {
+		s.scrapedMetricsPoints.Add(scraperCtx, int64(numScrapedMetrics), s.otelAttrs...)
+		s.erroredMetricsPoints.Add(scraperCtx, int64(numErroredMetrics), s.otelAttrs...)
+	} else { // OC for metrics
+		stats.Record(
+			scraperCtx,
+			obsmetrics.ScraperScrapedMetricPoints.M(int64(numScrapedMetrics)),
+			obsmetrics.ScraperErroredMetricPoints.M(int64(numErroredMetrics)))
+	}
+}
```
Comment (bogdandrutu) [inline obsreport/obsreport_scraper.go:190]:
```suggestion
}

```

File: obsreport/obsreporttest/otelprometheuschecker.go [hunk 7]
```diff
@@ -145,6 +152,12 @@ func fetchPrometheusMetrics(handler http.Handler) (map[string]*io_prometheus_cli
 	return parser.TextToMetricFamilies(rr.Body)
 }
 
+func attributesForScraperMetrics(receiver component.ID, scraper component.ID) []attribute.KeyValue {
+	return []attribute.KeyValue{
+		attribute.String(receiverTag.Name(), receiver.String()),
+		attribute.String(scraperTag.Name(), scraper.String()),
+	}
+}
```
Comment (bogdandrutu) [inline obsreport/obsreporttest/otelprometheuschecker.go:160]:
```suggestion
}

```

File: obsreport/obsreport_test.go [hunk 8]
```diff
@@ -219,62 +219,60 @@ func TestReceiveMetricsOp(t *testing.T) {
 }
 
 func TestScrapeMetricsDataOp(t *testing.T) {
-	tt, err := obsreporttest.SetupTelemetry()
-	require.NoError(t, err)
-	t.Cleanup(func() { require.NoError(t, tt.Shutdown(context.Background())) })
-
-	parentCtx, parentSpan := tt.TracerProvider.Tracer("test").Start(context.Background(), t.Name())
-	defer parentSpan.End()
+	testTelemetry(t, func(tt obsreporttest.TestTelemetry, registry *featuregate.Registry) {
```
Comment (bogdandrutu) [inline obsreport/obsreport_test.go:222]: This API is horrible, makes everything be super indented...

/cc @paivagustavo

**Review by bogdandrutu — APPROVED** [review]

File: obsreport/obsreport_scraper.go (hunk 6 above)
Comment (moh-osman3) [reply obsreport/obsreport_scraper.go]: Weird I thought I added this newline. Something in git must be preventing me from adding a newline at the end of the file (linter?)

File: obsreport/obsreport_test.go (hunk 8 above)
Comment (paivagustavo) [reply obsreport/obsreport_test.go:222]: @jmacd suggested the creation of new top level function tests as https://github.com/open-telemetry/opentelemetry-collector/pull/6423/commits/5961f605ec41cab61ac1abd1f75d6f3a3cc06657, which keeps the same indentation for the existing tests and can be easily removed once we're done with the transition.

cc @moh-osman3

**Review by bogdandrutu — APPROVED** [review]

File: obsreport/obsreport_scraper.go [hunk 9]
```diff
@@ -47,25 +66,67 @@ type ScraperSettings struct {
 
 // NewScraper creates a new Scraper.
 func NewScraper(cfg ScraperSettings) (*Scraper, error) {
-	return &Scraper{
+	return newScraper(cfg, featuregate.GetRegistry())
+}
+
+// Deprecated: [v0.65.0] use NewScraper.
+func MustNewScraper(cfg ScraperSettings) *Scraper {
```
Comment (codeboten) [inline obsreport/obsreport_scraper.go:73]: just clarifying this wasn't added here, the code moving in this PR makes it look like it was

Comment (moh-osman3) [reply obsreport/obsreport_scraper.go:73]: Unsure why its all highlighted green, something weird must have happened when rebasing. But my branch is up to date

File: obsreport/obsreport_test.go (hunk 8 above)
Comment (moh-osman3) [reply obsreport/obsreport_test.go]: Let me know if my latest commit matches what was suggested.

**Review by codeboten — APPROVED** [review]
Thanks @moh-osman3!

**Pull request merged**